File: elf.c
#include "common.h"
#include "task.h"

#pragma pack(push, 1)
struct elf64_header {
  uint32_t magic;
  uint8_t class;
  uint8_t endian;
  uint8_t version1;
  uint8_t abi_major;
  uint8_t abi_minor;
  uint32_t pad1;
  uint16_t pad2;
  uint8_t pad3;
  uint16_t type;
  uint16_t arch;
  uint32_t version2;
  uint64_t entrypoint;
  uint64_t ph_offset;
  uint64_t sh_offset;
  uint32_t flags;
  uint16_t mysize;
  uint16_t ph_entsize;
  uint16_t ph_entnum;
  uint16_t sh_entsize;
  uint16_t sh_entnum;
  uint16_t sh_section_name_entry_idx;
};

struct elf64_sectionheader {
  uint32_t name;
  uint32_t type;
  uint64_t flags;
  uint64_t addr;
  uint64_t offset;
  uint64_t size;
  uint32_t link;
  uint32_t info;
  uint64_t addralign;
  uint64_t entsize;
};
#pragma pack(pop)

int cor_elf_exec(char *elf, unsigned int len) {
  cor_printk("Called cor_elf_exec with elf=%x, len=%x\n",elf, len);
  if(sizeof(struct elf64_header) != 64) {
    cor_printk("assertion failed: elf64 header struct is size %x\n",sizeof(struct elf64_header));
    return -1;
  }

  if(len < sizeof(struct elf64_header)) {
    cor_printk("assertion failed: elf64 data too short at %x\n", len);
    return -1;
  }

  struct elf64_header *hdr = (struct elf64_header *)elf;

  if(0x464c457f != hdr->magic) {
    cor_printk("ERROR: magic of %x did not match ELF header of 0x464c457f\n", hdr->magic);
    return -1;
  } else {
    cor_printk("ELF magic looks OK.\n");
  }

  if(sizeof(struct elf64_sectionheader) != hdr->sh_entsize) {
    cor_printk("ERROR: ELF section header entry size invalid\n");
    return -1;
  }

  cor_printk("Nsections: %x\n", hdr->sh_entnum);

  if(hdr->sh_entnum > 30) {
    cor_printk("ELF: too many sections");
    return -1;
  }

  struct task_table_entry *t = task_new();

  for(int i = 0; i < hdr->sh_entnum; i++) {
    void *p = elf + hdr->sh_offset + hdr->sh_entsize * i;
    struct elf64_sectionheader *sectionheader = (struct elf64_sectionheader *) p;

    // TODO: is it allowed to have LMA=0 for non-debug-aux sections?
    if(sectionheader->addr == 0 || sectionheader->size == 0) {
      continue;
    }

    void *addr_space_start = (void*)sectionheader->addr;
    void *addr_space_end = (void*)sectionheader->addr + sectionheader->size;
    cor_printk("Section-used address space is: %p -- %p\n", addr_space_start, addr_space_end);

    void *lowpage = (void*)((uint64_t)addr_space_start & ~0xfff);
    void *highpage = (void*)ALIGN((uint64_t)addr_space_end, 0x1000);
    cor_printk("This means we need to make pages from %p -- %p\n", lowpage, highpage);

    for(void *page = lowpage; page < highpage; page += 0x1000) {
      cor_printk("Adding page at %p\n", page);

      // FIXME: we massively leak pages here by adding the same page multiple times
      task_addpage(t, page);
    }
  }



  task_enter_memspace(t);

  for(int i = 0; i < hdr->sh_entnum; i++) {
    void *p = elf + hdr->sh_offset + hdr->sh_entsize * i;
    struct elf64_sectionheader *sectionheader = (struct elf64_sectionheader *) p;

    // TODO: see above
    if(sectionheader->addr == 0 || sectionheader->size == 0) {
      continue;
    }

    //struct task_section *s = task_add_section(t, 1, sectionheader->size);
    char *source = elf + sectionheader->offset;

    cor_printk("section %x loading from %p to %p (size %x)\n",
      i, source, sectionheader->addr, sectionheader->size);

    for(size_t j = 0; j < sectionheader->size; j++) {
      char *loadtarget = (char*)sectionheader->addr + j;
      char *loadsrc = source + j;
      *loadtarget = *loadsrc;
    }
  }

  // memory sanity check; this is the "push rbp" opcode that should (always?) be the entry point
  // not really a good marker, but meh
  if(0x4855 != *((uint16_t *)(hdr->entrypoint))) {
    cor_printk("Virtual memory sanity check failed %x\n", *((uint16_t *)(hdr->entrypoint)));
    return -1;
  }

  void (*entry)() = (void(*)(void *))(hdr->entrypoint);
  cor_printk("entry = %x\n", entry);

  cor_printk("will now jump to %x\n",entry);

  int codeseg = 24;
  int segsel = codeseg | 3; // set RPL=3

  // switch segments for the call
  __asm__ (
    //"cli" // TODO later
    "pushq $35\n" // new SS, probably ignored
    "pushq $0x68000\n" // new RSP
    "pushf\n"
    "pushq %1\n" // new Code segment (important!)
    "pushq %0\n"
    "iretq\n"
  : : "entry" (entry), "segsel" (segsel) );

  cor_printk("done?!\n");
  return 0;
}

File: src/usertask/elf.rs
use core::{mem,slice};
use core::ptr::copy;
use core::iter::{Iterator,IntoIterator};

use super::super::mem::*;

#[derive(Debug)]
pub struct Image {
  pub initial_rsp: usize,
  pub initial_rip: usize,
}

#[derive(Debug)]
pub enum Error {
  InvalidElf
}

macro_rules! ensure {
  ($x:expr) => (if !$x { return Err(Error::InvalidElf)} );
}

type task = u8; // sigh

extern {
  fn task_new() -> *mut task;
  fn task_addpage(t: *mut task, page: *const task);
  fn task_enter_memspace(t: *mut task);
}

// Unsafe because we don't respect the address space at all.
// TODO: actually store which segments and pages we allocated.
pub unsafe fn load(elf: &[u8]) -> Result<Image, Error> {
  println!("Loading Elf from {:p}, len {}", elf.as_ptr(), elf.len());

  assert_eq!(64, mem::size_of::<Elf64Header>());
  ensure!(elf.len() >= mem::size_of::<Elf64Header>());

  let hdr: &Elf64Header = mem::transmute(elf.as_ptr());

  if hdr.magic != 0x464c457f {
    println!("ERROR: magic value {:x} did not match ELF header of 0x464c457f", hdr.magic);
    return Err(Error::InvalidElf);
  } else {
    println!("ELF magic looks OK.");
  }

  ensure!(hdr.sh_entsize as usize == mem::size_of::<Elf64SectionHeader>());

  let first_section_off = hdr.sh_offset as usize;
  let n_sections = hdr.sh_entnum as usize;

  ensure!(0 < n_sections);
  ensure!(n_sections < 30);

  let task = task_new();

  ensure!(elf.len() >= first_section_off
            + n_sections * mem::size_of::<Elf64SectionHeader>());

  let sections: &[Elf64SectionHeader] = unsafe {
    slice::from_raw_parts(mem::transmute(elf.as_ptr().offset(first_section_off as isize)), n_sections)
  };

  for s in sections.iter().filter(|s| s.addr != 0 && s.size != 0) {
    println!("Found a nonempty section: [{:x}; {:x}] {:?}", s.addr, s.size, s);

    let startpage = align_down(s.addr as usize, 0x1000); // align down
    let endpage = align_up((s.addr+s.size) as usize, 0x1000); // align up
    println!("It needs these pages: {:x} - {:x}", startpage, endpage);

    assert!(endpage > startpage); // empty sections are disallowed

    let mut page = startpage;
    while page < endpage {
      println!("Adding page at {:x}", page);

      // TODO: I believe task_addpage filters duplicates, but unsure
      task_addpage(task, page as *const u8);
      page += 0x1000;
    }
  }

  task_enter_memspace(task); // This is super unsafe!

  let mut brk = 0;

  for section in sections.iter().filter(|s| s.addr != 0 && s.size != 0) {
    if section._type == 0x8 && section.flags == 0x3 {
      // This seems to indicate the .data section, i.e. in the traditional memory layout,
      // the program break will be at the end of this section.
      // Our current, fairly ridiculous setup is to use this break location as the initial stack.
      // This means we will inevitably write into our own data. But we'll get to that later.

      // The least we can do is round up to the nearest page bound to cut us some slack.
      // (We allocated the entire page anyway in the above loop)
      brk = align_up((section.addr + section.size) as usize, 0x1000);
      println!("Found data section at {:x}, setting brk to {:x}", section.addr, brk);
    }

    ensure!(elf.len() >= (section.offset as usize) + (section.size as usize));
    let data_source = elf.as_ptr().offset(section.offset as isize) as *const u8;
    let data_dest = section.addr as *mut u8;

    copy(data_source, data_dest, section.size as usize);
  }

  // We should have found the break.
  ensure!(brk != 0);

  // Memory sanity check: This should be the opcode for "push %rbp", the first instruction
  // in _start. Actually a horrible way to check this, but meh.
  let firstword = hdr.entrypoint as *const u16;
  ensure!(*firstword == 0x4855);

  Ok(Image{initial_rip: hdr.entrypoint as usize, initial_rsp: brk})
}

#[derive(Debug)]
#[repr(C, packed)]
struct Elf64Header {
  magic: u32,
  class: u8,
  endian: u8,
  version1: u8,
  abi_major: u8,
  abi_minor: u8,
  pad1: u32,
  pad2: u16,
  pad3: u8,
  _type: u16,
  arch: u16,
  version2: u32,
  entrypoint: u64,
  ph_offset: u64,
  sh_offset: u64,
  flags: u32,
  mysize: u16,
  ph_entsize: u16,
  ph_entnum: u16,
  sh_entsize: u16,
  sh_entnum: u16,
  sh_section_name_entry_idx: u16,
}

#[derive(Debug)]
#[repr(C, packed)]
struct Elf64SectionHeader {
  name: u32,
  _type: u32,
  flags: u64,
  addr: u64,
  offset: u64,
  size: u64,
  link: u32,
  info: u32,
  addralign: u64,
  entsize: u64,
}
